### DISCUSSION TITLE
Which one is more stable AndroidJava or AndroidKotlin ?

### DISCUSSION THREAD

**Discussion by milansoliya4210** [WorldWindEarth/WorldWindKotlin#31] [Q&A] [answered]
Dear Team, can you suggest me which one is more stable as of now compare to AndroidJava or AndroidKotlin ?

**Comment by EMaksymenko** [comment, marked as answer]
Hi,

AndroidKotlin is more stable as AndroidJava is obsolete and not maintained
anymore.

пт, 30 черв. 2023, 07:41 користувач Milan Soliya ***@***.***>
пише:

**Comment by milansoliya4210** [comment]
If we want to implement 2D view then how it possible ?

**Comment by EMaksymenko** [comment]
For now it is not possible in Kotlin and Android version. Only JS and Java.

You may help to implement MercatorProjection class in
feature/mercator-projection fork.
This feature is half-implemented there.

пт, 30 черв. 2023, 11:11 користувач Milan Soliya ***@***.***>
пише:

**Comment by milansoliya4210** [comment]
can we integrate being map and ceisum terrain with this app using wms ?? and can we load other format apart from geopackage ?

**Comment by EMaksymenko** [comment]
You can use any WMS map or terrain server

пт, 30 черв. 2023, 12:39 користувач Milan Soliya ***@***.***>
пише:

**Comment by milansoliya4210** [comment]
geotiff ,ecw or .shp supported ?

**Comment by EMaksymenko** [comment]
Geotiff, bil16 and bil32 supported.

пт, 30 черв. 2023, 13:50 користувач Milan Soliya ***@***.***>
пише:

**Comment by milansoliya4210** [comment]
Hi @ComBatVision if I have to update position every two second and then the display view should be in center of location so for that what is ideal way ?? I am try to do wwd.engine.cameraFromLookAt(lookAt) once every position update but display view update after touch to screen, is there any way to sync view simultaneously ?

**Comment by EMaksymenko** [comment]
You need to call wwd.requestRedraw() to apply changes on UI.

вт, 4 лип. 2023, 12:35 користувач Milan Soliya ***@***.***>
пише:

**Reply by milansoliya4210** [reply]
Thanks @ComBatVision Worked

**Comment by milansoliya4210** [comment]
Hi, I want to measure distance from two points (contains lat, lon) how can I measure for WGS84 using this API ?  and  If I want to find point p2  which is at 100 m (suppose) from point 1 and bearing between point1 and point 2 is 350 then this possible with this API ?

**Comment by EMaksymenko** [comment]
Hello.

This can be done by methods getDistance, getAngle and calcPoint in Location
class, but there are three types of distances: great circle, rhumb or
linear, which leads to 3 sets of methods. You need to pick which one do you
need.

пт, 21 лип. 2023, 12:23 користувач Milan Soliya ***@***.***>
пише:

**Reply by milansoliya4210** [reply]
ok, thanks , if I have to consider ellipsoidal model (WGS84) for calculation then it is possible with this API?

**Comment by EMaksymenko** [comment]
Location calculates distance in radians. You need to multiply it by
elipsoid majorAxis.

пт, 21 лип. 2023, 14:16 користувач Milan Soliya ***@***.***>
пише:

**Comment by milansoliya4210** [comment]
Hi @ComBatVision  is there any way to show placemark with heading(azimuth wrt true north) like any shape or rotating icon same as google map has circle with orientation  ??

**Comment by EMaksymenko** [comment]
We are usually creating another placemark with arrow icon parallel to the terrain, because arrow in 3d space can not be shown on tactical symbol plain.
![Screenshot_20230811-132814_ComBat 4](https://github.com/WorldWindEarth/WorldWindKotlin/assets/12487306/c785cb53-1838-4e9e-80e6-f5ea2cbd1c2f)

**Comment by milansoliya4210** [comment]
@ComBatVision if I have to indicate azimuth (0-360) then I can rotate icon and its orientation would be not change in 3d  mode ??

**Comment by EMaksymenko** [comment]
Create another icon with tilt angle 90 degrees and rotate it in the plane of terrain surface

**Comment by milansoliya4210** [comment]
Thanks @ComBatVision  
 `  placeMark.imageRotationReference = OrientationMode.RELATIVE_TO_GLOBE`
`  placeMark.imageRotation = Angle.fromDegrees(teamNode.heading.toDouble())`

Simply Work for me

**Comment by milansoliya4210** [comment]
Hi @ComBatVision I have issue while render image tile from local storage , If I have data of zoom level 13 but for 12 not available then it  will load tiles for zoom level 13 ??
This is code for loading


`override fun getImageSource(x: Int, y: Int, z: Int): ImageSource {
        val file = File("$osmFolderPath/$z/$x/$y.png")
        return ImageSource.fromFile(file)
 }
`

**Comment by EMaksymenko** [comment]
If you have a zoom for level 12 but tiles only from 13 then it will not
show anything.

пт, 3 лист. 2023, 18:37 користувач Milan Soliya ***@***.***>
пише:

**Reply by milansoliya4210** [reply]
@ComBatVision currently what happen if I don't have data of 12 zoom level but i am try to zoom in to 13 so getImageSource(x: Int, y: Int, z: Int) try to call for zoom level 12  and it gone through exception so that it can not load zoom level 13 also

**Comment by EMaksymenko** [comment]
No, if it requires level 13 it will load level 13.

пт, 3 лист. 2023, 19:00 користувач Milan Soliya ***@***.***>
пише:

**Comment by EMaksymenko** [comment]
If you see stack trace of 12 level than system is still not require level
13. Zoom in more.

пт, 3 лист. 2023, 19:06 користувач Євген Максименко ***@***.***>
пише:

**Comment by milansoliya4210** [comment]
ok, @ComBatVision and if for current zoom level if all tiles are not available then it will visible only available tile for same zoom level ??

**Comment by EMaksymenko** [comment]
It will show only available tiles or part of top level tile from 1-11 level.

сб, 4 лист. 2023, 11:44 користувач Milan Soliya ***@***.***>
пише:

**Reply by EMaksymenko** [reply]
Could you give example file?

**Reply by milansoliya4210** [reply]
solved by handle null pointer exception

```kotlin
override fun getImageSource(x: Int, y: Int, z: Int): ImageSource? {
    Log.i("Map", "Z :$z X:$x Y:$y")

    val file = File("$osmFolderPath/$z/$x/$y.png")
    if (file.exists()) {
        return ImageSource.fromFile(file)
    }
    return null
}
